lib/realtime/timepoint_status.ex

```diff
@@ -23,26 +23,108 @@ defmodule Realtime.TimepointStatus do
             timepoint_status: timepoint_status()
           }
 
-  @spec timepoint_status([StopTime.t()], Stop.id()) :: timepoint_status() | nil
-  def timepoint_status(stop_times, stop_id) do
+  @spec timepoint_status([StopTime.t()], Stop.id(), {float(), float()}) ::
+          timepoint_status() | nil
+  def timepoint_status(stop_times, stop_id, latlon) do
+    # future_stop_times starts with the stop that has stop_id
     {past_stop_times, future_stop_times} = Enum.split_while(stop_times, &(&1.stop_id != stop_id))
 
     case Enum.find(future_stop_times, &StopTime.is_timepoint?(&1)) do
       %StopTime{timepoint_id: next_timepoint_id} ->
-        # past_count needs +1 for the step between the current timepoint and the first of the past stops
-        past_count = count_to_timepoint(Enum.reverse(past_stop_times)) + 1
-        future_count = count_to_timepoint(future_stop_times)
+        stops_until_timepoint = count_to_timepoint(future_stop_times)
+
+        # +1 for the step between the most recent stop and the next stop
+        stops_in_this_timepoint_segment =
+          count_to_timepoint(Enum.reverse(past_stop_times)) + 1 + stops_until_timepoint
+
+        previous_stop_time = List.last(past_stop_times)
+
+        fraction_until_next_stop =
+          1.0 -
+            fraction_between_stops(
+              previous_stop_time && previous_stop_time.stop_id,
+              stop_id,
+              latlon
+            )
 
         %{
           timepoint_id: next_timepoint_id,
-          fraction_until_timepoint: future_count / (future_count + past_count)
+          fraction_until_timepoint:
+            (stops_until_timepoint + fraction_until_next_stop) / stops_in_this_timepoint_segment
         }
 
       nil ->
         nil
     end
   end
 
+  @doc """
+  How far a bus is between two stops.
+  0.0 means the bus is still at the first stop.
+  1.0 means it's at the second stop.
+  0.5 means it's halfway between.
+  If either stop is missing lat lon data, default to being at the second stop.
+
+  Approximates the path between the stops as a straight line.
+  If the bus is not directly between the two stops,
+  measures from the closest point to the bus that is between the stops.
+  """
+  @spec fraction_between_stops(Stop.id() | nil, Stop.id() | nil, {float(), float()}) :: float()
+  def fraction_between_stops(start_id, finish_id, {lat, lon}) do
+    stop_fn = Application.get_env(:skate, :stop_fn, &Gtfs.stop/1)
+    start = start_id && stop_fn.(start_id)
```

Comment:
Hmm... on failure, `with` returns the value that made it fail, which here would be nil, instead of the `1.0` that we want. It'd also still require the nil checks on the latlons, which would get even wordier:
```
with
  %Stop{latitude: start_latitude, longitude: start_longitude}
  when not is_nil(start_latitude) and not is_nil(start_longitude)
  <- start_id && stop_fn.(start_id), …
```

I don't like the big stack of `&&`, and I would like a way around it, though. Other ideas?

Reply:
Ah yeah, interesting, I wasn't thinking about the wrinkle of `nil`s. 

My first thought in situations like this is typically to split out another function. Here's a stab at an approach. (I also pushed a branch named `mss-precision-with` if that's easier.) Take it or leave it. It probably doesn't reduce the line count. I think it might be a little bit clean and readable, but I'll let you decide if you think so or not (or want to change it in any way of course).

    @spec fraction_between_stops(point(), Stop.id() | nil, Stop.id() | nil) :: float()
    def fraction_between_stops(latlon, start_id, finish_id) do
      with {:ok, {start_latitude, start_longitude}} <- stop_lat_lon(start_id),
          {:ok, {finish_latitude, finish_longitude}} <- stop_lat_lon(finish_id) do
        # Treating the latlons as if they're cartesian coordinates works fine at this scale.
        # We don't need fancy great circle distance measurements around the globe
        # A degree of longitude is smaller than a degree of latitude,
        # but if the bus is directly between the stops, that has no effect at all
        # and if the bus is not on that straight line, it will have an acceptably small effect.
        latlon
        |> fraction_between_points(
          {start_latitude, start_longitude},
          {finish_latitude, finish_longitude}
        )
        |> clamp(0.0, 1.0)
      else
        _ ->
          1.0
      end
    end

    @spec stop_lat_lon(Stop.id() | nil) :: {:ok, point()} | :missing_stop | :missing_lat_lon
    def stop_lat_lon(nil), do: :missing_stop

    def stop_lat_lon(stop_id) do
      stop_fn = Application.get_env(:skate, :stop_fn, &Gtfs.stop/1)

      case stop_fn.(stop_id) do
        nil ->
          :missing_stop

        %Stop{latitude: nil} ->
          :missing_lat_lon

        %Stop{longitude: nil} ->
          :missing_lat_lon

        %Stop{latitude: latitude, longitude: longitude} ->
          {:ok, {latitude, longitude}}
      end
    end

Reply:
Ooh! The `with` is still a bit messy, but the separate `stop_latlon` function is a good idea. I've taken that and put it back in the if and I'm happy with it now.